Frontend Developer Love 2020 blogpost (#598)
ordina-jworks/ordina-jworks.github.io: _posts/2020-03-02-frontend-developer-love-2020.md

* [Practical a11y for web apps](#practical-a11y-for-web-apps-by-bob-bijvoet)
* [Micro-interactions with React Spring](#micro-interactions-with-react-spring-by-emma-bostian)
* [But, you're not Facebook](#but-youre-not-facebook-by-kitze)
* [Beats, rhymes and unit tests](#beats-rhymes-and-unit-tests-by-tony-edwards)
* [GraphQL without a database](#graphql-without-a-database-by-roy-derks)
* [DX is the new black. Learnings from using Nuxt and Storybook at scale](#dx-is-the-new-black-learnings-from-using-nuxt-and-storybook-at-scale-by-aurélie-violette)
* [Refactor your life](#refactor-your-life-by-noer-paanakker--sima-milli)
* [Blazor with WebAssembly](#blazor-with-webassembly-by-don-wibier)
* [Audio Streaming - Using WebRTC for building your own Voice AIs](#audio-streaming---using-webrtc-for-building-your-own-voice-ais-by-lee-boonstra)
* [The future of real-time, offline, data](#the-future-of-real-time-offline-data-by-nader-dabit)
* [The state of WebAssembly](#the-state-of-webassembly-by-sendil-kumarn)
* [Conclusion](#conclusion)

# Introduction

The Frontend Developer Love conference day was the first of 3 conference days that I was visiting in Amsterdam.

> While the other 2 days were focussed on VueJS, the first day was a day filled with more generic frontend topics.

Reviewer: Het woord "day" komt nogal veel voor en dat maakt het moeilijker om te lezen. Misschien is het beter om deze zin te wijzigen naar: 
```suggestion
While the last 2 days were focussed on VueJS, the first day was filled with more generic frontend topics.
```

In fact, it allows you to create a component based UI with a combination of C\#, HTML and CSS.
Both the client and server-side code are written in C\#.
This has one big benefit: you can share code and libraries between your front- and backend code.

A really basic example of Blazor code would be this:
```javascript
<h1>Counter</h1>
<p>Current count: @currentCount</p>
<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
	private int currentCount = 0;

	private void IncrementCount()
	{
		currentCount++;
	}
}
```

Those familiar with C# and mainly .Net Razor pages, will recognise the syntax.
Razor allows you to write both your HTML and C\# in the same file.

One of the interesting features of Blazor is the ability to compile it to WebAssembly.
In fact, your client-side C\# code is being run by WebAssembly in your browser.

In a Blazor project, you even have your client-side and server-side code right besides each other.
A typical Blazor project has the following structure: 

* /Client
* /Server
* /Shared

In this structure, your whole frontend is situated in `/Client` while all your backend code is situated in `/Server`.

> If you are in need of any code sharing, you can put it into `/Shared` and it will be availabled in both `/Client` and `/Server`.  

Reviewer 2: Typo: ... and it will be **available** in both ...

> Check out [Don's talk](https://www.youtube.com/watch?v=ZAFqw952GQM){:target="_blank" rel="noopener noreferrer"}, [Don's Twitter](https://twitter.com/donwibier){:target="_blank" rel="noopener noreferrer"} and [Blazor](https://dotnet.microsoft.com/apps/aspnet/web-apps/blazor){:target="_blank" rel="noopener noreferrer"}

# Audio Streaming - Using WebRTC for building your own Voice AIs, by [Lee Boonstra](https://twitter.com/ladysign){:target="_blank" rel="noopener noreferrer"}

Tools like the [Google Assistant](https://assistant.google.com/){:target="_blank" rel="noopener noreferrer"}, [Amazon Alexa](https://developer.amazon.com/en-US/alexa){:target="_blank" rel="noopener noreferrer"} and [Apple's Siri](https://www.apple.com/siri/){:target="_blank" rel="noopener noreferrer"} are becoming more popular.
The usage of voice assistants is becoming more mainstream as prices are dropping and they are becoming less of a gimmick and more of a tool to use during your everyday life.

And while there are lots of developer tools available to integrate your app with these particular voice assistants, this might not be your best course of action.

> You're limited to the technical requirements of these assistants, they might be overkill for your usecase or they might not fit for your enterprise usage.

Reviewer 2: Typo: ... your **use case** or ...

[WebRTC](https://webrtc.org/){:target="_blank" rel="noopener noreferrer"} is an open web standard and is available as a regular JavaScript API in all major browsers.
It allows for real-time communication (RTC) in the form of audio and video communication via direct peer-to-peer communication.
[RecordRTC](https://github.com/muaz-khan/RecordRTC){:target="_blank" rel="noopener noreferrer"} is a WebRTC JavaScript library for audio, video, screen and canvas recording.
By combining RecordRTC together with a tool such as [Dialogflow](https://dialogflow.com/){:target="_blank" rel="noopener noreferrer"} that can parse voice recordings to text, we can create our very own voice assistant.

Tools such as Dialogflow use machine learning to parse voice recordings to achieve intent matching.
An intent categorizes an end-user's intention for one conversation turn.
By trying to match phrases or parts of phrases, Dialogflow tries to classify the end-user expression to the best intent.
It then tries to parse the input with the help of the intent to extract information.

> For the example of weather forecase queries, if Dialogflow is able to match the queries to the forecast intent, it knows it can try to extract information such as time and location.

Reviewer 2: Typo: ... weather **forecast** queries, ...

AppSync is a managed service that uses GraphQL to make it easy for applications to get the data they need from multiple sources with the option to have real-time updates.
Amplify is a framework to build cloud-based full-stack serverless apps.
By combining the forces of these two tools, we can create real-time, offline data, especially if we combine it with GraphQL.

Your data model can be defined by a GraphQL schema.
In your application, you can use GraphQL subscriptions to have real-time updates for your application data.
There are options to get updates such as long polling, server sent events and web sockets.
It depends on your use case to choose what option is best for you.

But getting real-time updates also forces you to think about conflict detection and resolution.
Like what do you do when you get multiple updates at the same time?
What if your connection is down for a while?
There are a couple of popular ways to tackle these issues.

AppSync already uses the solution of monotonic counters combined with a base table that contains all your base data while also maintaining a change table to log all operations that happen on the base data.

> AppSync will automerge everything for you while also offering other options if necessary.

Reviewer 2: Typo: AppSync will **auto-merge/auto merge** everything ...

Check the [Amplify DataStore documentation](https://aws-amplify.github.io/docs/js/datastore){:target="_blank" rel="noopener noreferrer"} for more information.

> Check out [Nader's talk](https://www.youtube.com/watch?v=dkMEkD9OsPY){:target="_blank" rel="noopener noreferrer"}, [Nader's Twitter](https://twitter.com/dabit3){:target="_blank" rel="noopener noreferrer"}, [AWS AppSync](https://aws.amazon.com/appsync/){:target="_blank" rel="noopener noreferrer"}, [AWS Amplify](https://aws.amazon.com/amplify/){:target="_blank" rel="noopener noreferrer"} and the [Amplify DataStore documentation](https://aws-amplify.github.io/docs/js/datastore){:target="_blank" rel="noopener noreferrer"}

# The state of WebAssembly, by [Sendil Kumarn](https://twitter.com/sendilkumarn){:target="_blank" rel="noopener noreferrer"}
The final talk of the day was by Sendil who came to explain what the current state of WebAssembly is.

> WASM, short for WebAssembly, is a highlevel definition of how to run bytecode in your JavaScript engine.

Reviewer 2: Typo: ... is a **high-level/high level** definition of ...

> At its core, it's a stackmachine that uses a linear memory model by using a shared array buffer.

Reviewer 2: Typo: it's a **stack machine** that ...

If you're a fan of C/C++, the tool [Emscripten](https://github.com/emscripten-core){:target="_blank" rel="noopener noreferrer"} helps you out by compiling a C/C++ module to an HTML page.
For the Rust lovers, you can simply annotate a function with `#[wasm_bindgen]` to activate the [wasm-bindgen library](https://rustwasm.github.io/wasm-bindgen/){:target="_blank" rel="noopener noreferrer"} and allowing to compile a WASM function.
For the TypeScript lovers, there is a strict subset of TypeScript called [AssemblyScript](https://docs.assemblyscript.org/){:target="_blank" rel="noopener noreferrer"} that allows for compilation.

So the tools are already there to write WASM code but the future looks really bright as multiple features are currently being specified:

* [Interface types](https://github.com/WebAssembly/interface-types){:target="_blank" rel="noopener noreferrer"} to help describe higher-level values such as strings and records
* [Single Instruction Multiple Data](https://github.com/WebAssembly/simd){:target="_blank" rel="noopener noreferrer"} so that WASM will be more efficient on newer instruction set architectures up to 128-bit.
* Garbage collection
* ... and more!

> All this will lead to an even better performance of WebAssembly combined with more usecases so that WASM can become a more popular standard on the web.

Reviewer 2: Typo: ... more **use cases** so ...
